src/routes/Batch.tsx

```diff
@@ -0,0 +1,139 @@
+import React, { useEffect, useState, FormEvent } from "react";
+import { useSelector } from "react-redux";
+import { getBatchById } from "../store/batches/batches.selectors";
+import { useParams } from "react-router-dom";
+import { getProductById } from "../store/products/products.selectors";
+import {
+  useFetchBatchIngredients,
+  useAddBatchIngredient
+} from "../store/batch-ingredients/batch-ingredients.hooks";
+import { useFetchBatches } from "../store/batches/batches.hooks";
+import { useFetchProducts } from "../store/products/products.hooks";
+import { BatchIngredient } from "../store/batch-ingredients/batch-ingredients.types";
+import { getIngredientById } from "../store/ingredients/ingredients.selectors";
+import format from "date-fns/format";
+import { useFetchIngredients } from "../store/ingredients/ingredients.hooks";
+import IngredientSelect from "../components/IngredientSelect";
+import { IngredientId } from "../store/ingredients/ingredients.types";
+import { BatchId } from "../store/batches/batches.types";
+import ingredients from "../store/ingredients/ingredients.reducer";
+
+interface BatchIngredientListItemProps {
+  batchIngredient: BatchIngredient;
+}
+const BatchIngredientListItem: React.FC<BatchIngredientListItemProps> = props => {
+  const { batchIngredient } = props;
+  const ingredient = useSelector(
+    getIngredientById(batchIngredient.ingredient_id)
+  );
+  if (!ingredient) {
+    return null;
+  }
+
+  return (
+    <li>
+      {ingredient.name} ({ingredient.type}), {batchIngredient.quantity_kg}
+      {ingredient.unit}
+    </li>
+  );
+};
+
+interface AddBatchIngredientProps {
+  batch_id: BatchId;
+}
+const AddBatchIngredient: React.FC<AddBatchIngredientProps> = props => {
+  const { batch_id } = props;
+  const [ingredientId, setIngredientId] = useState<IngredientId | undefined>(undefined);
+  const [quantity, setQuantity] = useState<number | undefined>(undefined);
+
+  const { submit: fetchBatchIngredients } = useFetchBatchIngredients(batch_id);
+  const { submit: addBatchIngredient, status } = useAddBatchIngredient();
+
+  const ingredient = useSelector(getIngredientById(ingredientId || -1));
+
+  const onSubmit = (e: FormEvent) => {
+    e.preventDefault();
```

Comment:
is this a sort of type assignment of React type FormEvent to a variable `e'?

Reply:
when the form is submitted a "submit event" is fired
https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit_event

It might be there's actually a TypeScript `SubmitEvent` which i didn't know about, will check.
This function would be called the submit event handler. It is the function that handles the submit event. When a submit event is fired, the handler is called with the event has its argument. This is an html/javscript phenomenon and not specific to typescript/react.

`e.preventDefault()` prevents the default handling of the event. In this case, when i form is submitted the default way it is handled (by the browser) is to take the user to another page, or reload the page. We don't want that though, we just want to fire off an xhr request. So we must prevent the default action.

https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault

Reply:
Got it, :)